ring-core/src/ring/middleware/file.clj

```diff
@@ -40,14 +40,20 @@
   Accepts the following options:
 
   :index-files?    - look for index.* files in directories, defaults to true
-  :allow-symlinks? - serve files through symbolic links, defaults to false"
+  :allow-symlinks? - serve files through symbolic links, defaults to false
+  :handler-first? - let handler process the files first. if handler successfully processes the file, do not serve static content"
   ([handler root-path]
    (wrap-file handler root-path {}))
   ([handler root-path options]
    (ensure-dir root-path)
    (fn
      ([request]
-      (or (file-request request root-path options) (handler request)))
+      (if (:handler-first? options)
+        (let [{:keys [status] :as response-map} (handler request)]
+          (if (= status 200)
```

Comment:
This should be:

    (if (not= (:status response) 404)

Reply:
what if the status  is 400? could we check to make sure the status code is not 400-499 ? it seems those numbers are reserved for error status

Reply:
If the status is 400 then we should return the response from the handler. The only time we don't is when the handler tells us unambiguously that it cannot find the resource (a 404). All other responses, including error responses, should be returned unaltered.